File: problems/2696. Minimum String Length After Removing Substrings.py
# link: https://leetcode.com/problems/minimum-string-length-after-removing-substrings/

# Stack
class Solution:
    def minLength(self, s: str) -> int:
        stack = []
        char_map = {'B': 'A', 'D': 'C'}
        length = len(s)
        for c in s:
            match c:
                case 'A' | 'C':
                    stack.append(c)
                case c if c in char_map and stack and stack[-1] == char_map[c]:
                    stack.pop()
                    length -= 2
                case _:
                    stack = []

        return length

File: problems/2696. Minimum String Length After Removing Substrings.cpp
// link: https://leetcode.com/problems/minimum-string-length-after-removing-substrings/

// Modify In Place
#include<map>
#include <string>
using std::string;

class Solution {
public:
    int minLength(string s) {
        int writePtr = 0;
        std::map<char, char> charMap;
        charMap['B'] = 'A';
        charMap['D'] = 'C';

        for (int readPtr=0; readPtr < s.length(); readPtr++) {
            s[writePtr] = s[readPtr];

            if (writePtr > 0 &&
             charMap.find(s[writePtr]) != charMap.end() &&
             s[writePtr-1] == charMap[s[writePtr]]) {
                writePtr--;
             } else {
                writePtr++;
             }
        }
        return writePtr;
    }
};
